### PR TITLE
Feat/icq delegation chunks

### PR DESCRIPTION
**Pull request by ratik**
[Task](https://hadronlabs.atlassian.net/browse/DRP-246)

### CODE REVIEW

File: contracts/puppeteer/src/contract.rs [hunk 1]
```diff
@@ -153,26 +140,67 @@ fn query_kv_query_ids(deps: Deps<NeutronQuery>) -> ContractResult<Binary> {
...
+                delegations: vec![],
+            },
+            remote_height: 0,
+            local_height: 0,
+            timestamp: Timestamp::default(),
+        }),
+        Some(last_key) => {
+            let last_data = puppeteer_base
+                .delegations_and_balances
+                .may_load(deps.storage, &last_key)?
```
Comment (foxpy) [inline contracts/puppeteer/src/contract.rs:159]: Why is it may_load here? If there is a key for data, shouldn't we expect this data to be always there?

File: contracts/puppeteer/src/contract.rs [hunk 2]
```diff
@@ -153,26 +140,67 @@ fn query_kv_query_ids(deps: Deps<NeutronQuery>) -> ContractResult<Binary> {
...
+        None => to_json_binary(&BalancesResponse {
+            balances: Balances { coins: vec![] },
+            remote_height: 0,
+            local_height: 0,
+            timestamp: Timestamp::default(),
+        }),
+        Some(last_key) => {
+            let last_data = puppeteer_base
+                .delegations_and_balances
+                .may_load(deps.storage, &last_key)?
```
Comment (foxpy) [inline contracts/puppeteer/src/contract.rs:189]: and same here

File: contracts/puppeteer/src/contract.rs (hunk 1 above)
Comment (ratik) [reply contracts/puppeteer/src/contract.rs:159]: initial state, when we don't have any complete set of answers

Comment (ratik) [reply contracts/puppeteer/src/contract.rs:159]: but in general it seems we can just `load` here. wdyt?

Comment (foxpy) [reply contracts/puppeteer/src/contract.rs:159]: yes, that's what I am thinking, just using load() should be sufficient

File: integration_tests/src/testcases/puppeteer.test.ts [hunk 3]
```diff
@@ -102,6 +108,7 @@ describe('Interchain puppeteer', () => {
   });
 
   it('instantiate', async () => {
+    await context.park.restartRelayer('neutron', 1);
```
Comment (foxpy) [inline integration_tests/src/testcases/puppeteer.test.ts:111]: is it a debugging artifact or does it actually provide any value?

File: contracts/puppeteer/src/contract.rs (hunk 1 above)
Comment (ratik) [reply contracts/puppeteer/src/contract.rs:159]: well, from core pow, we need these queries work even at start wo any complete set or even delegations. This way let's keep may_load with returning empty struct

File: integration_tests/src/testcases/puppeteer.test.ts [hunk 4]
```diff
@@ -141,6 +148,7 @@ describe('Interchain puppeteer', () => {
         account.address,
         res.codeId,
         {
+          delegations_queries_chunk_size: 2,
```
Comment (foxpy) [inline integration_tests/src/testcases/puppeteer.test.ts:151]:
```suggestion
          delegations_queries_chunk_size: VALIDATORS_ICQ_LIMIT,
```

File: integration_tests/src/testcases/puppeteer.test.ts (hunk 3 above)
Comment (ratik) [reply integration_tests/src/testcases/puppeteer.test.ts:111]: when a query is deleted, the relayer in most cases exits with the exception about submission data for non-existent query

**Review by foxpy — CHANGES_REQUESTED** [review]

File: integration_tests/src/testcases/puppeteer.test.ts (hunk 3 above)
Comment (foxpy) [reply integration_tests/src/testcases/puppeteer.test.ts:111]: Yes, but you are triggering a restart in a test called 'instantiate', which is ordered before any other test which might delete queries.

File: contracts/puppeteer/src/contract.rs (hunk 1 above)
Comment (foxpy) [reply contracts/puppeteer/src/contract.rs:159]: You are using .ok_or(Error) right at the next line anyway :^)
Might as well just switch to using load()

File: packages/puppeteer-base/src/reply.rs [hunk 5]
```diff
@@ -28,6 +28,20 @@ where
...
     }
 
+    pub fn register_delegations_and_balance_query_reply(
+        &self,
+        deps: DepsMut,
+        msg: Reply,
+        i: u16,
+        query_type: U,
+    ) -> StdResult<Response> {
+        let query_id = get_query_id(msg.result.clone())?;
```
Comment (foxpy) [inline packages/puppeteer-base/src/reply.rs:38]: I am certain you don't need this clone() here

File: integration_tests/src/testcases/puppeteer.test.ts (hunk 3 above)
Comment (ratik) [reply integration_tests/src/testcases/puppeteer.test.ts:111]: oh my bad 👍

File: packages/puppeteer-base/src/sudo.rs [hunk 6]
```diff
@@ -94,6 +96,62 @@ where
...
+        deps: DepsMut<NeutronQuery>,
+        env: Env,
+        query_id: u64,
+        version: &str,
+    ) -> NeutronResult<Response<NeutronMsg>> {
+        let chunks_len = self
+            .delegations_and_balances_query_id_chunk
+            .keys(deps.storage, None, None, Order::Ascending)
+            .collect::<Vec<_>>()
+            .len();
```
Comment (foxpy) [inline packages/puppeteer-base/src/sudo.rs:110]:
```suggestion
            .count();
```

File: packages/puppeteer-base/src/sudo.rs [hunk 7]
```diff
@@ -94,6 +96,62 @@ where
...
+            Some(mut state) => {
+                state
+                    .data
+                    .delegations
+                    .delegations
+                    .extend(data.delegations.delegations);
+                // insert chunk_id to collected_chunks and uniq
+                state.collected_chunks.push(chunk_id);
+                state.collected_chunks.sort();
+                state.collected_chunks.dedup();
```
Comment (foxpy) [inline packages/puppeteer-base/src/sudo.rs:135]: I see you dedup collected_chunks and I agree with that. Don't you also have to dedup delegations, or I am being wrong about that?

File: packages/puppeteer-base/src/sudo.rs [hunk 8]
```diff
@@ -94,6 +96,62 @@ where
...
+                remote_height,
+                local_height: env.block.height,
+                timestamp: env.block.time,
+                collected_chunks: vec![chunk_id],
+            },
+        };
+        if new_state.collected_chunks.len() == chunks_len {
+            self.last_complete_delegations_and_balances_key
+                .save(deps.storage, &env.block.height)?;
+        }
```
Comment (foxpy) [inline packages/puppeteer-base/src/sudo.rs:149]: Before overriding last complete key, please check that the key you save is actually newer than previous one.

**Review by foxpy — CHANGES_REQUESTED** [review]

File: packages/puppeteer-base/src/sudo.rs (hunk 8 above)
Comment (ratik) [reply packages/puppeteer-base/src/sudo.rs:149]: can you imagine `env.block.height` is less than for the any previous tx?

File: packages/puppeteer-base/src/sudo.rs (hunk 7 above)
Comment (ratik) [reply packages/puppeteer-base/src/sudo.rs:134]: the validators in chunks are separated and merging answers for the chunks is no other than the prev logic with having only one ICQ

File: packages/puppeteer-base/src/sudo.rs (hunk 8 above)
Comment (foxpy) [reply packages/puppeteer-base/src/sudo.rs:148]: uhhh, don't you think we should group chunks by remote height and not by local height?

Comment (ratik) [reply packages/puppeteer-base/src/sudo.rs:148]: ok, valid 👍

File: packages/puppeteer-base/src/sudo.rs (hunk 7 above)
Comment (foxpy) [reply packages/puppeteer-base/src/sudo.rs:134]: Well, if a chunk gets pushed twice, then state.data.delegations would have duplicate data, and we don't want that to happen. Maybe instead we should simply do nothing when relayer submits a chunk that we already have?

**Review by foxpy — APPROVED** [review]

**Review by albertandrejev — APPROVED** [review]

**Pull request merged**